src/Param.cc

```diff
@@ -427,6 +427,84 @@ bool ParseColorUsingStringStream(const std::string &_input,
   return isValidColor;
 }
 
+//////////////////////////////////////////////////
+/// \brief Helper function for Param::ValueFromString for parsing pose
+/// This checks the pose components specified in _input are xyzrpy or xyzwxyz
+/// (expects 6 ir 7 values) and the resulting pose is valid.
+/// \param[in] _input Input string.
+/// \param[in] _key Key of the parameter, used for error message.
+/// \param[out] _value This will be set with the parsed value.
+/// \return True if parsing pose succeeded.
+bool ParsePoseUsingStringStream(const std::string &_input,
+    const std::string &_key, ParamPrivate::ParamVariant &_value)
+{
+  StringStreamClassicLocale ss(_input);
+  std::string token;
+  std::vector<double> values;
+  double v;
+  bool isValidPose = true;
+  while (ss >> token)
+  {
+    try
+    {
+      v = std::stod(token);
+    }
+    // Catch invalid argument exception from std::stod
+    catch(std::invalid_argument &)
+    {
+      sdferr << "Invalid argument. Unable to set value ["<< token
+             << "] for key [" << _key << "].\n";
+      isValidPose = false;
+      break;
+    }
+    // Catch out of range exception from std::stof
+    catch(std::out_of_range &)
+    {
+      sdferr << "Out of range. Unable to set value [" << token
+             << "] for key [" << _key << "].\n";
+      isValidPose = false;
+      break;
+    }
+
+    if (values.size() > 7)
+    {
+      sdferr << "Pose values cannot have more than 7 values.\n";
+      isValidPose = false;
+      break;
+    }
+
+    if (!std::isfinite(v))
+    {
+      sdferr << "Pose values must be finite.\n";
+      isValidPose = false;
+      break;
+    }
+
+    values.push_back(v);
+  }
+
+  if (!isValidPose)
+    return ParseUsingStringStream<ignition::math::Pose3d>(_input, _key, _value);
+
+  if (values.size() == 6u)
+  {
+    _value = ignition::math::Pose3d(values[0], values[1], values[2],
+        values[3], values[4], values[5]);
+  }
+  else if (values.size() == 7u)
+  {
+    _value = ignition::math::Pose3d(values[0], values[1], values[2],
```

Comment:
I played around with this idea, but since we are dealing with string parsing, the string comparisons will have to happen somewhere else too, if not here, in order to figure out which representation `enum` it is.

I considered doing adding something like this to `Types.hh`,
```cpp
  /// \brief Enum of supported rotation types,
  enum class RotationType : std::size_t
  {
    RpyRadians = 0,
    RpyDegrees = 1,
    Quaternion = 2
  };

  /// \brief Description of a rotation by it's name and number of elements.
  using RotationDescription = std::pair<const char*, std::size_t>;

  /// \brief Map of rotation type to rotation descriptions.
  using RotationDescriptionMap = std::array<RotationDescription, 3>;

  /// \brief Map of supported rotation types to their descriptions.
  constexpr RotationDescriptionMap kRotationTypes = {{
    {"rpy_radians", 6},
    {"rpy_degrees", 6},
    {"q_wxyz", 7}
  }};
```

There'll still be a need to compare to use `IGN_DTOR`, which doesn't really justify adding anything to the public API. Keeping these comparisons in implementation might be the cleanest way to do it.

Reply:
This looks better! I have some other comments here, but will make them as new discussions.